### PR TITLE
feat: hardware wallet transactions

### PR DESCRIPTION
**Pull request by SWvheerden**
Description
---
Intial draft for hardware wallet transactions. 

Motivation and Context
---

How Has This Been Tested?
---

### CODE REVIEW

File: src/RFC-0205_LedgerTransactions.md [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+The helper can attach the correct signatures to the [UTXO]s and ship the transaction.
+
+### Receiving normal one-sided transaction
+This can be done by the helper asking the signer for a public key. And advertising this public key as the destination public key for a 1-sided transaction. 
+The helper can scan the blockchain for this public key. 
+
+### Receiving one-sided-stealth
+`TODO`
+
+### Output recovery
```
Comment (AaronFeickert) [inline src/RFC-0205_LedgerTransactions.md:124]: It may be useful to describe how the helper can check the validity of an output's commitment and script public key after decrypting `v` and `k_i`:
- The helper checks that `v` and `k_i` are a valid opening of the commitment by reconstructing it
- The helper then checks that `K_s` corresponds to `k_i` and `A`

This process can detect malformed outputs that would not be spendable.

**Review by AaronFeickert — CHANGES_REQUESTED** [review]

File: src/RFC-0205_LedgerTransactions.md [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+Because the key (\\( k_H \\) ) is calculated from the seed phrase of the signer, this will be the same each time. The helper can try to decrypt each scanned output, when it is successful it knows it has found its own output. 
+
+## Security
+Because the script key is required for spending, it is the only key that needs to be kept secret. 
+The following table explains what an attacker can do upon learning a key from the helper
+
+| key         | Worst case scenario          | 
+|:------------|:-----------------------------|
+| \\( k_H \\) | Can view all transactions and values made by the wallet |
+| \\( k_i \\) | Can try and brute force the value of the transaction |
```
Comment (AaronFeickert) [inline src/RFC-0205_LedgerTransactions.md:135]: This can also occur if `K_i` were leaked.

**Comment by AaronFeickert** [comment]
What should be the relationship between this RFC and the existing [RFC-0204](https://rfc.tari.com/RFC-0204_TariScriptExamples.html) (which discusses interactive and "legacy" one-sided outputs) and [RFC-0203](https://rfc.tari.com/RFC-0203_StealthAddresses.html) (which discusses stealth one-sided outputs)?

File: src/RFC-0205_LedgerTransactions.md [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+## Background
+
+Vanilla [Mimblewimble] only has a single secret per [UTXO], the blinding factor (\\( k_i \\) ), the Tari protocol extends the [Mimblewimble] protocol to include scripting in the form of [TariScript]. This adds a second secret per [UTXO] called the script_key (\\( k_s \\) ). In practical terms this means that while vanilla [Mimblewimble] only requires that a wallet wishing to spend an [UTXO], prove knowledge of (\\( k_i \\) ) by singing the kernel signature, this is not sufficient for Tari. A Tari wallet must also prove knowledge of the script key (\\( k_s \\) ), by signing the script signature. 
+
+## Requirements
+
+To properly implement hardware wallets we need the following requirements to be met:
+* No UTXO can be spent without a user physically approving the transaction on the hardware wallet.
+* Users need to verify transaction properties when signing for transactions.
+* The user must be able to receive transactions without having to authorize them on the hardware wallet.
```
[the remark this answers is not in the document]
Comment (SWvheerden) [reply src/RFC-0205_LedgerTransactions.md:79]: Umm, with interactive payments, we cannot have a case of a `malicious sender effectively burning funds`. 
The receiver side if wholly constructed by the receiver. 
A malicious receiver is a whole different ball game and I don't think possible to stop

File: src/RFC-0205_LedgerTransactions.md [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+All [UTXO]s created will be created with a script `PushPubkey(K_S)`. The key (\\( k_s \\) ) is created as follows:
+$$
+\begin{aligned}
+s = Hash(k_i * A) \\\\
+k_S = s + a \\\\
+K_S = s \cdot G + A
+\end{aligned}
+$$
+
+The blinding factor (\\( k_i \\) ) is used as a random nonce when creating the script key. This means the helper can create the public key without the signer present, and the signer can then at a later stage create the private key from the nonce. The key pair (\\( a, A \\) ) is the master private key from the signer. The private key (\\( a \\) ) is kept secret by the signer at all times.
```
[the remark this answers is not in the document]
Comment (SWvheerden) [reply src/RFC-0205_LedgerTransactions.md:100]: just made it key pair?

File: src/RFC-0205_LedgerTransactions.md (hunk 2 above)
Comment (SWvheerden) [reply src/RFC-0205_LedgerTransactions.md:135]: Yeah this table is here for that regardless of how the key is leaked or learned

File: src/RFC-0205_LedgerTransactions.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (AaronFeickert) [reply src/RFC-0205_LedgerTransactions.md:79]: Isn't it still possible for a sender to duplicate an existing valid value-and-mask ciphertext with a malformed commitment? Then if the scanner is only defining "receiving" as being able to decrypt the ciphertext, it would be fooled and claim the recipient has funds that in fact cannot be spent.

File: src/RFC-0205_LedgerTransactions.md [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+* Signer: This is the device that keeps the secrets for the transactions and approves them, aka the hardware wallet.
+* Helper: This device is the program that helps the signer construct the transaction and send it over the network, aka wallet.
+
+### Process Overview
+By splitting the ownership of the [UTXO]'s secret by assigning knowledge of only the script key (\\( k_s \\) ) to the signer, we can lift much of the heavy cryptography like bulletproof creation to the helper device by exposing (\\( k_i \\) ) to it. By looking at how [one-sided-stealth] transactions are created, we can construct the script key in such a way that the helper can calculate the public script key, but cannot calculate the private script key.
+
+All [UTXO]s created will be created with a script `PushPubkey(K_S)`. The key (\\( k_s \\) ) is created as follows:
+$$
+\begin{aligned}
+s = Hash(k_i * A) \\\\
```
[the remark this answers is not in the document]
Comment (AaronFeickert) [reply src/RFC-0205_LedgerTransactions.md]: This should be something akin to `H(k_i)`, not just `k_i` itself.

File: src/RFC-0205_LedgerTransactions.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (SWvheerden) [reply src/RFC-0205_LedgerTransactions.md:79]: The sender does not construct this, the receiver does. So how can the sender force the receiver to malform its own commitment?

File: src/RFC-0205_LedgerTransactions.md (hunk 2 above)
Comment (AaronFeickert) [reply src/RFC-0205_LedgerTransactions.md:135]: Sure, just noting that you don't need `k_i` to do the brute forcing of the value, as `K_i` is sufficient for it too.

File: src/RFC-0205_LedgerTransactions.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (AaronFeickert) [reply src/RFC-0205_LedgerTransactions.md:79]: Looks like you already made an update later in the RFC to address this. Apologies!

File: src/RFC-0205_LedgerTransactions.md [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+The helper can attach the correct signatures to the [UTXO]s and ship the transaction.
+
+### Receiving normal one-sided transaction
+This can be done by the helper asking the signer for a public key. And advertising this public key as the destination public key for a 1-sided transaction. 
+The helper can scan the blockchain for this public key. 
+
+### Receiving one-sided-stealth
+`TODO`
+
+### Output recovery
```
Comment (AaronFeickert) [inline src/RFC-0205_LedgerTransactions.md:123]: Perhaps worth noting that this process only works for interactive outputs.

For legacy one-sided outputs, the helper can watch for `A` in scripts (as you note elsewhere), but since it can't complete the DHKE required to decrypt the commitment opening, it can't assert the output is valid and must send it to the signer to be checked.

For stealth one-sided outputs, the helper can't do anything except send such outputs to the signer to be checked.

File: src/RFC-0205_LedgerTransactions.md [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+
+All [UTXO]s created will be created with a script `PushPubkey(K_S)`. The key (\\( k_s \\) ) is created as follows:
+$$
+\begin{aligned}
+k_S = k_i + a \\\\
+K_S = k_i \cdot G + A
+\end{aligned}
+$$
+
+The blinding factor (\\( k_i \\) ) is used as a random nonce when creating the script key. This means the helper can create the public key without the signer present, and the signer can then at a later stage create the private key from the nonce. The key pair (\\( a, A \\) ) is the master key pair from the signer. The private key (\\( a \\) ) is kept secret by the signer at all times.
```
Comment (CjS77) [inline src/RFC-0205_LedgerTransactions.md:99]: When you say "master key pair", does this mean that the same `a` is used for every transaction?

File: src/RFC-0205_LedgerTransactions.md [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+
+### Initialization
+Adding a hardware wallet to a wallet we need to ensure that all keys are only derived from a single seed phrase provided by the hardware wallet. 
+
+Helper asks signer for master helper key (\\( k_H \\) ). 
+This key is derived from the signer seed phrase.
+
+### Transaction receiving
+When a transaction is received the helper constructs the new [UTXO] with its Rangeproof. Choosing a new (\\( k_i \\) ) for the [UTXO], it calculates a new \\( s \\) and \\( K_S \\). It attaches the script `PushPubkey(K_S)` to output.
```
Comment (CjS77) [inline src/RFC-0205_LedgerTransactions.md:108]: What is `s` here?

File: src/RFC-0205_LedgerTransactions.md [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+
+### Transaction sending
+When the user wants to send a transaction, the helper retrieves the desired [UTXO]. The helper gives asks the signer to sign the transaction.
```
Comment (CjS77) [inline src/RFC-0205_LedgerTransactions.md:111]:
```suggestion
When the user wants to send a transaction, the helper retrieves the desired [UTXO]. The helper asks the signer to sign the transaction. 
```

File: src/RFC-0205_LedgerTransactions.md [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+
+### Entities
+Normal transactions have only a single entity, the wallet which controls all secrets and transactions. But with hardware wallets, we need to define two distinct types:
+* Signer: This is the entity that keeps the secrets for the transactions and approves them, aka the hardware wallet.
+* Helper: This entity is the program that helps the signer construct the transaction, send it over the network and scan the network, aka wallet.
+
+### Process Overview
+By splitting the ownership of the [UTXO]'s secret by assigning knowledge of only the script key (\\( k_s \\) ) to the signer, we can lift much of the heavy cryptography like bulletproof creation to the helper device by exposing (\\( k_i \\) ) to it. By looking at how [one-sided-stealth] transactions are created, we can construct the script key in such a way that the helper can calculate the public script key, but cannot calculate the private script key.
+
+All [UTXO]s created will be created with a script `PushPubkey(K_S)`. The key (\\( k_s \\) ) is created as follows:
```
Comment (CjS77) [inline src/RFC-0205_LedgerTransactions.md:91]: I gather, from the discussion below, that you mean to say that we must update the protocol to deprecate the current `NOP` default script so that there's always a script key present?

File: src/RFC-0205_LedgerTransactions.md [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+### Receiving normal one-sided transaction
+This can be done by the helper asking the signer for a public key. And advertising this public key as the destination public key for a 1-sided transaction. 
+The helper can scan the blockchain for this public key. 
+
+### Receiving one-sided-stealth
+`TODO`
+
+### Output recovery
+When creating outputs the wallet encrypts the blinding factor (\\k_i \\) and value \\( v \\) with (\\( k_H \\) ).
+Because the key (\\( k_H \\) ) is calculated from the seed phrase of the signer, this will be the same each time. The helper can try to decrypt each scanned output, when it is successful it knows it has found its own output.
```
Comment (CjS77) [inline src/RFC-0205_LedgerTransactions.md:125]: Are `a` and `k_H` the same thing?

Comment (CjS77) [inline src/RFC-0205_LedgerTransactions.md:125]: So it's a bit worrying to use the same key every time from a security perspective.

Could we perhaps use a strategy similar to the one in [BIP-32](https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki#user-content-Child_key_derivation_CKD_functions) for deriving an infinite number of `A'` from a single parent private key (signer) or the parent public key (helper)?

**Review by CjS77 — COMMENTED** [review]
I like it. Some discussion points and clarifying questions.

**Review by CjS77 — COMMENTED** [review]
I like it. Some discussion points and clarifying questions.

File: src/RFC-0205_LedgerTransactions.md (hunk 6 above)
Comment (AaronFeickert) [reply src/RFC-0205_LedgerTransactions.md:123]: Yeah, for legacy one-sided it's mostly a UX issue, and somewhat a "wait for the signer" issue. For stealth one-sided it's an entirely different problem that does not scale.

File: src/RFC-0205_LedgerTransactions.md (hunk 7 above)
Comment (AaronFeickert) [reply src/RFC-0205_LedgerTransactions.md:99]: Yes, it would be.

File: src/RFC-0205_LedgerTransactions.md (hunk 11 above)
Comment (AaronFeickert) [reply src/RFC-0205_LedgerTransactions.md:125]: No, `k_H` is a fixed helper key derived from `a` that's used for encrypting commitment openings. The signer provides it once to the helper, who can then use it for AEAD operations.

Comment (AaronFeickert) [reply src/RFC-0205_LedgerTransactions.md:125]: There's no security problem here. Encrypted openings use an extended nonce whose entire purpose is to allow for safe key reuse with random nonce sampling. And this key reuse allows for super-fast interactive output scanning during recovery.

Comment (AaronFeickert) [reply src/RFC-0205_LedgerTransactions.md:125]: Also, the wallet public key `A` doesn't play any role in AEAD decryption during recovery. It only comes into play when testing for offset script public keys, and in that case, there's already a hash-based offset applied that masks the fixed `A`.

File: src/RFC-0205_LedgerTransactions.md (hunk 8 above)
Comment (SWvheerden) [reply src/RFC-0205_LedgerTransactions.md:108]: was removed in a refactor, fixed

File: src/RFC-0205_LedgerTransactions.md [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+
+### Transaction sending
+When the user wants to send a transaction, the helper retrieves the desired [UTXO]. The helper asks the signer to sign the transaction. 
+The singer calculates \\( k_s \\) to sign the transaction. 
+The signer creates a random nonce \\( k_O \\) to use for the script_offset. It produces the metadata signature with \\( k_O \\), and supplies the script_offset to the helper. 
+The helper can attach the correct signatures to the [UTXO]s and ship the transaction.
+
+### Receiving normal one-sided transaction
+This can be done by the helper asking the signer for a public key. And advertising this public key as the destination public key for a 1-sided transaction. 
+The helper can scan the blockchain for this public key.
```
Comment (stringhandler) [inline src/RFC-0205_LedgerTransactions.md:118]: What happens to the wallet's Tari Address? Is this not related?

File: src/RFC-0205_LedgerTransactions.md [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+The singer calculates \\( k_s \\) to sign the transaction. 
+The signer creates a random nonce \\( k_O \\) to use for the script_offset. It produces the metadata signature with \\( k_O \\), and supplies the script_offset to the helper. 
+The helper can attach the correct signatures to the [UTXO]s and ship the transaction.
+
+### Receiving normal one-sided transaction
+This can be done by the helper asking the signer for a public key. And advertising this public key as the destination public key for a 1-sided transaction. 
+The helper can scan the blockchain for this public key. 
+
+### Receiving one-sided-stealth
+`TODO`
```
Comment (stringhandler) [inline src/RFC-0205_LedgerTransactions.md:121]: Just adding a marker here. DNM until resolved

**Review by stringhandler — COMMENTED** [review]
I think it's worth adding a section on generic TariScripts that are not one-sided, e.g. HTLC

File: src/RFC-0205_LedgerTransactions.md [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+the host machines that run the wallets are compromised by malware the secrets from the transactions are still safe as they are not
+kept on the machine as with regular wallets. The devices are very low power and only feature a very limited processing power. MimbleWimble 
+has intensive crypto operations that are in some instances very slow or not able to run on the hardware wallet at all. This RFC describes a way to 
+get around these limitations to have fully functional secure hardware wallet integration. 
+
+## Background
+
+Vanilla [Mimblewimble] only has a single secret per [UTXO], the blinding factor (\\( k_i \\) ), the Tari protocol extends the [Mimblewimble] protocol to include scripting in the form of [TariScript]. This adds a second secret per [UTXO] called the script_key (\\( k_s \\) ). In practical terms this means that while vanilla [Mimblewimble] only requires that a wallet wishing to spend an [UTXO], prove knowledge of (\\( k_i \\) ) by producing the kernel signature, this is not sufficient for Tari. A Tari wallet must also prove knowledge of the script key (\\( k_s \\) ), by producing the script signature.
```
[the remark this answers is not in the document]
Comment (SWvheerden) [reply src/RFC-0205_LedgerTransactions.md:72]: Practically No, Technically yes. 
A `No-Op` script that just contains the `No-Op` operation accepts any input. 

So from the wallet submitting the tx, it can submit any Public key as input, and it's still valid. But it still needs to prove it knows the secret key of that public key.  

But from the prover side, aka base_node, it just runs, the script and verifies the signature, thus proving knowledge of the script key k_S. How the key was calculated from the script the Base_node does not care.

File: src/RFC-0205_LedgerTransactions.md [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+
+## Background
+
+Vanilla [Mimblewimble] only has a single secret per [UTXO], the blinding factor (\\( k_i \\) ), the Tari protocol extends the [Mimblewimble] protocol to include scripting in the form of [TariScript]. This adds a second secret per [UTXO] called the script_key (\\( k_s \\) ). In practical terms this means that while vanilla [Mimblewimble] only requires that a wallet wishing to spend an [UTXO], prove knowledge of (\\( k_i \\) ) by producing the kernel signature, this is not sufficient for Tari. A Tari wallet must also prove knowledge of the script key (\\( k_s \\) ), by producing the script signature. 
+
+## Requirements
+
+To properly implement hardware wallets we need the following requirements to be met:
+* No UTXO can be spent without a user physically approving the transaction on the hardware wallet.
+* Users need to verify transaction properties when signing for transactions.
```
[the remark this answers is not in the document]
Comment (SWvheerden) [reply src/RFC-0205_LedgerTransactions.md:78]: yes

File: src/RFC-0205_LedgerTransactions.md (hunk 12 above)
Comment (SWvheerden) [reply src/RFC-0205_LedgerTransactions.md:118]: Currently, the wallet calculates the Public script key for the script from the wallet's Tari Address. The only reason that is used is that it's an easily known key that the wallet has the secret key to. So we use that key in place of the TariAddress

File: src/RFC-0205_LedgerTransactions.md (hunk 7 above)
Comment (AaronFeickert) [reply src/RFC-0205_LedgerTransactions.md:99]: It's faster, and there's no need to do ECDH for this use case.

File: src/RFC-0205_HardwareTransactions.md [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+
+To properly implement hardware wallets we need the following requirements to be met:
+* No UTXO can be spent without a user physically approving the transaction on the hardware wallet.
+* Users need to verify transaction properties when signing for transactions.
+* The user must be able to receive transactions without having to authorize them on the hardware wallet.
+
+## Implementation
+
+### Entities
+Normal transactions have only a single entity, the wallet which controls all secrets and transactions. But with hardware wallets, we need to define two distinct types:
```
[the remark this answers is not in the document]
Comment (AaronFeickert) [reply src/RFC-0205_HardwareTransactions.md:84]: And I'd use the term "entities" rather than "types" here, for consistency.

File: src/RFC-0205_HardwareTransactions.md [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+* Signer: This is the entity that keeps the secrets for the transactions and approves them, aka the hardware wallet.
+* Helper: This entity is the program that helps the signer construct the transaction, send it over the network and scan the network, aka wallet.
+
+### Process Overview
+By splitting the ownership of the [UTXO]'s secret by assigning knowledge of only the script key (\\( k_s \\) ) to the signer, we can lift much of the heavy cryptography like bulletproof creation to the helper device by exposing (\\( k_i \\) ) to it. By looking at how [one-sided-stealth] transactions are created, we can construct the script key in such a way that the helper can calculate the public script key, but cannot calculate the private script key.
```
Comment (AaronFeickert) [inline src/RFC-0205_HardwareTransactions.md:89]: Some context on what a UTXO's "secret" is could be helpful here. Really it's two secret values: the commitment mask and the script private key. (The value is private too, but can be easily brute forced and therefore not as relevant.)

File: src/RFC-0205_HardwareTransactions.md [hunk 18]
```diff
@@ -0,0 +1,148 @@
...
+
+All hardware wallet created [UTXO]s created will be created with a script `PushPubkey(K_S)`. The key (\\( k_s \\) ) is created as follows:
+$$
+\begin{align}
+k_S &= k_i + a \\\\
```
Comment (AaronFeickert) [inline src/RFC-0205_HardwareTransactions.md:94]: Still recommend not using the mask directly in this way.

File: src/RFC-0205_HardwareTransactions.md [hunk 19]
```diff
@@ -0,0 +1,148 @@
...
+K_S &= k_i \cdot G + A
```
Comment (AaronFeickert) [inline src/RFC-0205_HardwareTransactions.md:95]: Nor here.

File: src/RFC-0205_HardwareTransactions.md (hunk 18 above)
Comment (AaronFeickert) [reply src/RFC-0205_HardwareTransactions.md:94]: If this is for interactive outputs, I'd use an approach like `H(k_i) + a`. If the design choice is made to use the helper key `k_H` for the commitment opening encryption, the helper won't need to compute the corresponding script public key `H(k_i)*G + A` except on outputs it already flags via opening decryption.

File: src/RFC-0205_HardwareTransactions.md (hunk 17 above)
Comment (SWvheerden) [reply src/RFC-0205_HardwareTransactions.md:89]: It is discussed two paragraphs up?

File: src/RFC-0205_HardwareTransactions.md (hunk 18 above)
Comment (SWvheerden) [reply src/RFC-0205_HardwareTransactions.md:94]: Its going to have to compute `K_S` for every commitment it makes, it needs to put that into the script

Comment (AaronFeickert) [reply src/RFC-0205_HardwareTransactions.md:94]: Right; I meant for scanning purposes (but wasn't clear about that).

Comment (AaronFeickert) [reply src/RFC-0205_HardwareTransactions.md:94]: A similar attack could be used in an attempt to generically link transactions without knowledge of `A`.

**Review by AaronFeickert — CHANGES_REQUESTED** [review]
There are a few spots where the math-mode separators are written incorrectly, meaning not all math renders properly.

File: src/RFC-0205_HardwareTransactions.md [hunk 20]
```diff
@@ -0,0 +1,147 @@
...
+the host machines that run the wallets are compromised by malware the secrets from the transactions are still safe as they are not
+kept on the machine as with regular wallets. The devices are very low power and only feature a very limited processing power. MimbleWimble 
+has intensive crypto operations that are in some instances very slow or not able to run on the hardware wallet at all. This RFC describes a way to 
+get around these limitations to have fully functional secure hardware wallet integration. 
+
+## Background
+
+Vanilla [Mimblewimble] only has a single secret per [UTXO], the blinding factor (\\( k_i \\) ), the Tari protocol extends the [Mimblewimble] protocol to include scripting in the form of [TariScript]. This adds a second secret per [UTXO] called the script_key (\\( k_s \\) ). In practical terms this means that while vanilla [Mimblewimble] only requires that a wallet wishing to spend an [UTXO], prove knowledge of (\\( k_i \\) ) by producing the kernel signature, this is not sufficient for Tari. A Tari wallet must also prove knowledge of the script key (\\( k_s \\) ), by producing the script signature.
```
Comment (AaronFeickert) [inline src/RFC-0205_HardwareTransactions.md:71]: The first sentence is a run-on.

File: src/RFC-0205_HardwareTransactions.md [hunk 21]
```diff
@@ -0,0 +1,147 @@
...
+The signer calculates \\( k_s \\) to sign the transaction. 
+The signer creates a random nonce \\( k_O \\) to use for the script_offset. It produces the metadata signature with \\( k_O \\), and supplies the script_offset to the helper. 
+The helper can attach the correct signatures to the [UTXO]s and ship the transaction.
+
+### Receiving normal one-sided transaction
+This can be done by the helper asking the signer for a public key. And advertising this public key as the destination public key for a 1-sided transaction. 
+The helper can scan the blockchain for this public key. 
+
+### Receiving one-sided-stealth
+`TODO`
```
Comment (AaronFeickert) [inline src/RFC-0205_HardwareTransactions.md:120]: Not possible yet with this design.

File: src/RFC-0205_HardwareTransactions.md [hunk 22]
```diff
@@ -0,0 +1,147 @@
...
+
+### Output recovery
+When creating outputs the wallet encrypts the blinding factor \\(k_i \\) and value \\( v \\) with \\( k_H \\).
```
Comment (AaronFeickert) [inline src/RFC-0205_HardwareTransactions.md:123]: Might be worth noting that the encryption is done in practice using an extended-nonce AEAD, which allows for a random nonce and uses authenticated decryption for this detection process.

**Review by AaronFeickert — CHANGES_REQUESTED** [review]

File: src/RFC-0205_HardwareTransactions.md [hunk 23]
```diff
@@ -0,0 +1,147 @@
...
+
+### Entities
+Normal transactions have only a single entity, the wallet which controls all secrets and transactions. But with hardware wallets, we need to define two distinct entities:
+* Signer: This is the entity that keeps the secrets for the transactions and approves them, aka the hardware wallet.
+* Helper: This entity is the program that helps the signer construct the transaction, send it over the network and scan the network, aka wallet.
+
+### Process Overview
+By splitting the ownership of the [UTXO]'s secrets by assigning knowledge of only the script key (\\( k_s \\) ) to the signer, we can lift much of the heavy cryptography like bulletproof creation to the helper device by exposing (\\( k_i \\) ) to it. By looking at how [one-sided-stealth] transactions are created, we can construct the script key in such a way that the helper can calculate the public script key, but cannot calculate the private script key.
+
+All hardware wallet created [UTXO]s created will be created with a script `PushPubkey(K_S)`. The key (\\( k_s \\) ) is created as follows:
```
Comment (brianp) [inline src/RFC-0205_HardwareTransactions.md:90]: That's a lot of created.

```suggestion
All hardware wallet created [UTXO]s will be contain a script `PushPubkey(K_S)`. The key (\\( k_s \\) ) is created as follows:
```

**Review by CjS77 — APPROVED** [review]
